Sming/SmingCore/Network/WebsocketFrame.cpp

```diff
@@ -0,0 +1,253 @@
+
+	// create header
+
+	// byte 0
+	if (fin)
+	{
+		*_header |= bit(7);    ///< set Fin
+	}
+	*_header |= (uint8_t) opcode;        ///< set opcode
+	_header++;
+
+	// byte 1
+	if (mask)
+	{
+		*_header |= bit(7);    ///< set mask
+	}
+
+	if (length < 126)
+	{
+		*_header |= length;
+		_header++;
+	}
+	else if (length < 0xFFFF)
+	{
+		*_header |= 126;
+		_header++;
+		*_header = ((length >> 8) & 0xFF);
+		_header++;
+		*_header = (length & 0xFF);
+		_header++;
+	}
+
+	if (mask && headerToPayload)
+	{
+		//we work on copy of original data so we can mask it without affecting original
+
+		for (uint8_t x = 0; x < sizeof(maskKey); x++)
+		{
+			maskKey[x] = random(0xFF);
+			*_header = maskKey[x];
+			_header++;
+		}
+
+		uint8_t * dataMaskPtr = (_payload + _headerLength);
+
+		for (size_t x = 0; x < length; x++)
+		{
+			dataMaskPtr[x] = (dataMaskPtr[x] ^ maskKey[x % 4]);
+		}
+
+	}
+
+	if (headerToPayload)
+	{
+		_header = nullptr; //mark _header as nullptr to indicate external world that whole frame is referenced by _payload
+		_headerLength = 0;
+	}
+	else
+	{
+		_header = headerPtr;
+	}
+
+	return true;
+}
+
+uint8_t WebsocketFrameClass::_getFrameSizes(uint8_t* buffer, size_t length)
+{
+//	for (uint8_t x = 0; x < length - _nextReadOffset; x++)
+//	{
+//		debugf("buffer[%d] = 0x%x", x, buffer[x]);
+//	}
+
+	uint16_t payloadLength = buffer[1] & 0b01111111; // length of payload
+	_mask = buffer[1] & 0b10000000; // extracting Mask bit
+
+	//calculate header size
+	payloadLength < 126 ? _headerLength = 2 : _headerLength = 4;
+
+	if (_mask)
+	{
+		_headerLength += 4;
+	}
+
+	if (payloadLength == 126 && length < 4)
+	{
+		_frameType = WSFrameType::incomplete;
+		return false;
+	}
+
+	if (payloadLength == 127) //unsupported frame size!
+	{
+		_frameType = WSFrameType::error;
+		return false;
+	}
+
+	if (payloadLength == 126)
+	{
+		//next 2 bytes are length
+		payloadLength = buffer[2];
+		payloadLength << 8;
+		payloadLength |= buffer[3];
+	}
+
+	_payloadLength = payloadLength;
+
+	_nextReadOffset += _headerLength + _payloadLength; // if given buffer is multiframe buffer store read offset for next frame
+//	debugf("_headerLength: %d, _payloadLength: %d, _nextReadOffset: %d length: %d",_headerLength, _payloadLength, _nextReadOffset,length);
+
+	if (length == _nextReadOffset) // check if given buffer is single frame buffer
+	{
+		_nextReadOffset = 0; // single websocket frame in buffer
+	}
+
+	return true;
+}
+
+uint8_t WebsocketFrameClass::decodeFrame(uint8_t * buffer, size_t length)
+{
+	buffer += _nextReadOffset; // if called again for multiframe buffer, rewind buffer first to next frame
+									// no effect for single frame buffer as initial value of _nextReadOffset = 0
+
+	uint8_t op = buffer[0] & 0b00001111; // Extracting Opcode
+	uint8_t fin = buffer[0] & 0b10000000; // Extracting Fin Bit (Single Frame)
+
+	// debugf("Opcode : 0x%x",op);
+	// debugf("Fin : 0x%x",fin);
+	if (op == 0x00 || op == 0x01 || op == 0x02) //Data frames
```

Comment:
Can you remove the hard-coded values and replace them with easier to read OPCODE_XYZ... defines or WSFrameType enums?